File: plugins/plugin-discord/typescript/actions/listChannels.ts
import type {
  Action,
  ActionExample,
  ActionResult,
  Content,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { DISCORD_SERVICE_NAME } from "../constants";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import type { DiscordService } from "../service";

const spec = requireActionSpec("LIST_CHANNELS");

export const listChannels: Action = {
  name: spec.name,
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,
  validate: async (_runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    return message.content.source === "discord";
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult | undefined> => {
    const discordService = runtime.getService(DISCORD_SERVICE_NAME) as DiscordService;

    if (!discordService || !discordService.client) {
      runtime.logger.error(
        {
          src: "plugin:discord:action:list-channels",
          agentId: runtime.agentId,
        },
        "Discord service not found or not initialized"
      );
      return { success: false, error: "Discord service not available" };
    }

    try {
      // Get all allowed channels
      const allowedChannelIds = discordService.getAllowedChannels();

      if (allowedChannelIds.length === 0) {
        if (callback) {
          await callback?.({
            text: "I'm currently listening to all channels (no restrictions are set).",
            source: "discord",
          });
        }
        return { success: true, text: "Listening to all channels (no restrictions)" };
      }

      // Fetch channel information for each allowed channel
      const channelInfoPromises = allowedChannelIds.map(async (channelId) => {
        try {
          const client = discordService.client;
          const channel = client && (await client.channels.fetch(channelId));
          if (channel?.isTextBased() && !channel.isVoiceBased()) {
            const guild = "guild" in channel ? channel.guild : null;
            return {
              id: channelId,
              name: "name" in channel ? channel.name : "DM",
              mention: `<#${channelId}>`,
              server: guild?.name || "Direct Message",
            };
          }
        } catch (_e) {
          // Channel might have been deleted or bot lost access
          return {
            id: channelId,
            name: "Unknown",
            mention: channelId,
            server: "Unknown or Deleted",
          };
        }
        return null;
      });

      const channelInfos = (await Promise.all(channelInfoPromises)).filter(Boolean);

      // Format the response
      let responseText = `I'm currently listening to ${channelInfos.length} channel${channelInfos.length !== 1 ? "s" : ""}:\n\n`;

      // Group by server
      const channelsByServer = channelInfos.reduce(
        (acc, channel) => {
          if (!channel) {
            return acc;
          }
          if (!acc[channel.server]) {
            acc[channel.server] = [];
          }
          acc[channel.server].push(channel);
          return acc;
        },
        {} as Record<string, typeof channelInfos>
      );

      // Format by server
      for (const [serverName, channels] of Object.entries(channelsByServer)) {
        responseText += `**${serverName}**\n`;
        for (const channel of channels) {
          if (channel) {
            responseText += `• ${channel.name} (${channel.mention})\n`;
          }
        }
        responseText += "\n";
      }

      // Check if CHANNEL_IDS is set
      const envChannelIds = runtime.getSetting("CHANNEL_IDS") as string;
      if (envChannelIds) {
        responseText +=
          "\n*Some channels are configured in environment settings and cannot be removed dynamically.*";
      }

      const response: Content = {
        text: responseText.trim(),
        actions: ["LIST_CHANNELS_RESPONSE"],
        source: message.content.source,
      };

      if (callback) {
        await callback?.(response);
      }
      return { success: true, text: response.text };
    } catch (error) {
      runtime.logger.error(
        {
          src: "plugin:discord:action:list-channels",
          agentId: runtime.agentId,
          error: error instanceof Error ? error.message : String(error),
        },
        "Error listing channels"
      );
      if (callback) {
        await callback?.({
          text: "I encountered an error while trying to list the channels. Please try again.",
          source: "discord",
        });
      }
      return { success: false, error: error instanceof Error ? error.message : String(error) };
    }
  },
  examples: (spec.examples ?? []) as ActionExample[][],
};

export default listChannels;

File: plugins/plugin-discord/rust/src/actions/list_channels.rs
//! List channels action

use async_trait::async_trait;

use super::{ActionContext, ActionResult, DiscordAction};
use crate::error::Result;
use crate::DiscordService;

/// Action to list all channels the bot is listening to
pub struct ListChannelsAction;

#[async_trait]
impl DiscordAction for ListChannelsAction {
    fn name(&self) -> &str {
        "LIST_CHANNELS"
    }

    fn description(&self) -> &str {
        "Lists all Discord channels the bot is currently listening to and responding in."
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "SHOW_CHANNELS",
            "LIST_LISTENING_CHANNELS",
            "SHOW_MONITORED_CHANNELS",
            "GET_CHANNELS",
            "WHICH_CHANNELS",
            "CHANNELS_LIST",
        ]
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        let source = context
            .message
            .get("source")
            .and_then(|v| v.as_str())
            .unwrap_or("");
        Ok(source == "discord")
    }

    async fn handler(
        &self,
        _context: &ActionContext,
        service: &DiscordService,
    ) -> Result<ActionResult> {
        // Get all allowed channels
        let allowed_channel_ids = service.get_allowed_channels();

        if allowed_channel_ids.is_empty() {
            return Ok(ActionResult::success_with_data(
                "I'm currently listening to all channels (no restrictions are set).",
                serde_json::json!({
                    "channels": [],
                    "unrestricted": true,
                }),
            ));
        }

        // Fetch channel information
        let mut channel_infos = Vec::new();
        for channel_id in &allowed_channel_ids {
            match service.get_channel_info(channel_id).await {
                Ok(Some(info)) => channel_infos.push(info),
                _ => {
                    channel_infos.push(serde_json::json!({
                        "id": channel_id,
                        "name": "Unknown",
                        "server": "Unknown or Deleted",
                    }));
                }
            }
        }

        // Group by server
        let mut channels_by_server: std::collections::HashMap<String, Vec<serde_json::Value>> =
            std::collections::HashMap::new();

        for channel in &channel_infos {
            let server = channel
                .get("server")
                .and_then(|s| s.as_str())
                .unwrap_or("Unknown")
                .to_string();
            channels_by_server
                .entry(server)
                .or_default()
                .push(channel.clone());
        }

        // Format response
        let mut response_lines = vec![
            format!(
                "I'm currently listening to {} channel{}:",
                channel_infos.len(),
                if channel_infos.len() != 1 { "s" } else { "" }
            ),
            String::new(),
        ];

        for (server_name, channels) in &channels_by_server {
            response_lines.push(format!("**{}**", server_name));
            for channel in channels {
                let name = channel
                    .get("name")
                    .and_then(|n| n.as_str())
                    .unwrap_or("Unknown");
                let mention = channel
                    .get("mention")
                    .or_else(|| channel.get("id"))
                    .and_then(|m| m.as_str())
                    .unwrap_or("");
                response_lines.push(format!("• {} ({})", name, mention));
            }
            response_lines.push(String::new());
        }

        if service.has_env_channels() {
            response_lines.push(
                "*Some channels are configured in environment settings \
                and cannot be removed dynamically.*"
                    .to_string(),
            );
        }

        Ok(ActionResult::success_with_data(
            response_lines.join("\n").trim().to_string(),
            serde_json::json!({
                "channels": channel_infos,
                "count": channel_infos.len(),
            }),
        ))
    }
}
